Update ably-go to 1.2.0

tomczoink opened the pull request:
- At its core it has the ability to subscribe and publish to messages
- Also retrieves history when connecting, and presence events are published

cruickshankpg commented on chat.go:
```diff
@@ -0,0 +1,108 @@
+		AuthOptions: ably.AuthOptions{
+			// If you have an Ably account, you can find
+			// your API key at https://www.ably.io/accounts/any/apps/any/app_keys
+			Key: "INSERT_YOUR_API_KEY_HERE",
```
We shouldn't encourage people to hard code the api key. It should be an env var or a cli param.We use github.com/urfave/cli/v2 for our tools.

Jmgr replied:
I that worth it for a demo though? Could be using a const string at the top of the file maybe?

cruickshankpg replied:
If we tell customers not to do it then we shouldn't do it. Doing it as an env var is really simple and a cli is pretty straightforward.
```
Key: os.GetEnv("ABLY_KEY")
```

Jmgr replied:
Ah yes if we say that then it makes sense.

Jmgr commented on README.md:
```diff
@@ -0,0 +1,19 @@
+
+Next, replace the **INSERT_YOUR_API_KEY_HERE** line in `chat.go` with your Ably API key. You can sign up for an account with [Ably](https://www.ably.com/) and access your API key from the [app dashboard](https://www.ably.com/accounts/any/apps/any/app_keys).
+
+Finally, run the application with the following line, specify your ClientID as when asked, and you're ready to start communicating over Ably!
```
I think `as when asked` => `when asked`.

cruickshankpg commented on chat.go:
```diff
@@ -0,0 +1,108 @@
+func main() {
+	fmt.Println("Type your clientID")
+	reader := bufio.NewReader(os.Stdin)
+	clientId, _ := reader.ReadString('\n')
```
personal preference but make it a command arg. Also might be better to call it a username if this is a chat app.

Jmgr commented on chat.go:
```diff
@@ -0,0 +1,108 @@
+	fmt.Println("Type your clientID")
+	reader := bufio.NewReader(os.Stdin)
+	clientId, _ := reader.ReadString('\n')
+	clientId = strings.Replace(clientId, "\n", "", -1)
```
`strings.Replace(clientId, "\n", "", -1)` => `strings.ReplaceAll(clientId, "\n", "")`

Jmgr commented on chat.go:
```diff
@@ -0,0 +1,108 @@
+
+	for {
+		text, _ := reader.ReadString('\n')
+		text = strings.Replace(text, "\n", "", -1)
```
Same thing here with `ReplaceAll`.

Jmgr reviewed (commented):
I don't have too much experience with the Ably go client library, but LGTM :+1:

cruickshankpg commented on chat.go:
```diff
@@ -0,0 +1,108 @@
+func subscribeToPresence(channel ably.RealtimeChannel) {
+	sub, err := channel.Presence.Subscribe()
+	if err != nil {
+		panic(err)
```
panicking on an error is a bit messy. It would be nicer to return errors, print the error nicely for the user and exit.
I would use an golang.org/x/sync/errgroup for this. Probably no need to use a context in this case.

cruickshankpg commented on README.md:
````diff
@@ -0,0 +1,19 @@
+Firstly install the Ably Go library:
+
+```bash
+~ $ go get -u github.com/ably/ably-go/ably
````
not needed if you've set up a go.mod

cruickshankpg reviewed (changes requested):
You should set up a go.mod which will save the `go get` step and is standard go practice these days.
Run `go mod init` to set this up

cruickshankpg commented on chat.go:
```diff
@@ -0,0 +1,108 @@
+
+	getHistory(*channel)
+
+	go subscribeToPresence(*channel)
```
Minor: You also don't need separate go routines here for presence and channels you can do
``` go
sub, err := channel.Subscribe()
if err != nil {
    return fmt.Errorf("subscribing to channel: %w"; err)
}
pSub, err := channel.Presence.Subscribe()
if err != nil {
    return fmt.Errorf("subscribing to presence: %w"; err)
}
select {
case msg := <-sub.MessageChannel():
    ...
case msg := <- pSub.PresenceChannel():
    ....
}
```

lmars commented:
@tomczoink lot's of comments already so I'll hold off reviewing, but it's probably best to use the ably-go 1.2 preview which is soon to be released and includes a breaking API change (otherwise the demo is going to become stale over night).

The 1.2 preview is currently tagged `v1.2.0-apipreview.3` and you can see the new API in the README:

https://github.com/ably/ably-go/tree/v1.2.0-apipreview.3

Once you've initialised the modules (i.e. using `go mod init`), then update `ably-go` to use the 1.2 preview with:

```
go get github.com/ably/ably-go@v1.2.0-apipreview.3
```

tomczoink committed:
Update basic got chat app per feedback

- Uses newest Ably Go library
-- This involves various code changes to be compatable
- Add go.mod to simplify getting started
- Removed API key from core code for security

sacOO7 commented:
We don't need `ably-chat` binary file right?

tomczoink committed:
Delete ably-chat binary file

sacOO7 commented:
I feel you might need to upgrade preview to the latest v4 `v1.2.0-apipreview.4`

tomczoink committed:
Remove unused ClientOptions

tomczoink committed:
Update to ably-go 1.2.0
